Prevent Inconsistent State From Being Observed Outside STM Transaction

adamgfraser opened the pull request:
Resolves #6168.

Even when the result of an STM transaction is not successful we still have to confirm that the journal is valid with the global lock before executing the to dos.

mijicd reviewed (approved):
It would be good to include this issue in the STM test suite or integrate the JCStress test @durban provided. Other than that, looks good to me!

adamgfraser commented:
Added a stress test. Also identified and fixed a regression where interruption could be observed after a transaction had been successfully committed.

Yes, we should definitely fix on `master` as well but I thought given the differences in the STM implementations made sense to do here first and then we can back port.
